### ISSUE TITLE
Flatten Function Missing in 0.10.5

### ISSUE THREAD

**Issue by osterman** [hashicorp/terraform#16115]
### Terraform Version
`0.10.5`

### Terraform Configuration Files
```hcl
resource "null_resource" "test" {
  triggers {
    value = "${flatten(list(list(1,2,3)))}"
  }
}

output "test" {
  value = "${null_resource.test.triggers.value}"
}
```

### Debug Output
```
localhost:/conf/erik/test# terraform plan
1 error(s) occurred:

* null_resource.test: 1 error(s) occurred:

* null_resource.test: 1:3: unknown function called: flatten in:

${flatten(list(list(1,2,3)))}
```

### Expected Behavior

Per the [documentation](https://www.terraform.io/docs/configuration/interpolation.html#flatten-list-of-lists-) we expected the nested lists to be flattened. 

### Actual Behavior

Terraform exits with an error: `unknown function called: flatten`

### Steps to Reproduce
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`

**Comment by apparentlymart** [comment]
Hi @osterman! Sorry this didn't work as expected.

Thanks for providing this great minimal reproduction case. When I tried to run it I ran into a couple different errors:

* The inner `list` call wants its functions to be strings. (This actually seems like a bug, but something that should get addressed as part of forthcoming config language improvements where we're going to revamp the type system.)
* The result of the whole expression must be a string to assign to a triggers value.

So I updated the config as follows:

```hcl
resource "null_resource" "test" {
  triggers {
    value = "${join(",", flatten(list(list("1","2","3"))))}"
  }
}

output "test" {
  value = "${null_resource.test.triggers.value}"
}
```

After this, I was able to apply it:

```
$ terraform apply
null_resource.test: Creating...
  triggers.%:     "" => "1"
  triggers.value: "" => "1,2,3"
null_resource.test: Creation complete after 0s (ID: 4064527121356582727)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

test = 1,2,3
```

So with that said I'm not sure what's going on here. :thinking:

Is it possible that the `terraform` binary you're running isn't the version you think it is? The `flatten` function is pretty new (added in 0.10.3) so this result would be expected for a version prior to that.
